packages/core/contracts/Escrow.sol

```diff
@@ -171,283 +166,337 @@ contract Escrow is IEscrow, ReentrancyGuard {
         status = EscrowStatuses.Pending;
 
         remainingFunds = getBalance();
-        require(remainingFunds > 0, 'Escrow balance is zero');
+        require(remainingFunds > 0, 'Zero balance');
 
         emit PendingV2(
-            manifestUrl,
-            manifestHash,
-            reputationOracle,
-            recordingOracle,
-            exchangeOracle
+            _url,
+            _hash,
+            _reputationOracle,
+            _recordingOracle,
+            _exchangeOracle
         );
         emit Fund(remainingFunds);
     }
 
+    /**
+     * @dev Cancels the escrow and transfers remaining funds to the canceler.
+     * @return bool indicating success of the cancellation.
+     */
     function cancel()
         public
         override
-        trusted
+        adminOrLauncher
         notBroke
-        notComplete
         nonReentrant
         returns (bool)
     {
-        _safeTransfer(token, canceler, remainingFunds);
-        status = EscrowStatuses.Cancelled;
-        remainingFunds = 0;
-        emit Cancelled();
+        status = EscrowStatuses.ToCancel;
+        emit CancellationRequested();
         return true;
     }
 
+    /**
+     * @dev Withdraws excess funds from the escrow for a specific token.
+     * @param _token Address of the token to withdraw.
+     * @return bool indicating success of the withdrawal.
+     */
     function withdraw(
         address _token
-    ) public override trusted nonReentrant returns (bool) {
-        uint256 _amount;
+    ) public override adminOrLauncher nonReentrant returns (bool) {
+        uint256 amount;
         if (_token == token) {
-            uint256 _balance = getBalance();
-            require(_balance > remainingFunds, 'No funds to withdraw');
-            _amount = _balance - remainingFunds;
+            uint256 balance = getBalance();
+            require(balance > remainingFunds, 'No funds');
+            amount = balance - remainingFunds;
         } else {
-            _amount = getTokenBalance(_token);
+            amount = getTokenBalance(_token);
         }
 
-        _safeTransfer(_token, canceler, _amount);
+        require(amount > 0, 'No funds');
+        IERC20(_token).safeTransfer(msg.sender, amount);
 
-        emit Withdraw(_token, _amount);
+        emit Withdraw(_token, amount);
         return true;
     }
 
-    function complete() external override notExpired trustedOrReputationOracle {
+    /**
+     * @dev Completes the escrow, transferring remaining funds to the launcher.
+     */
+    function complete() external override notExpired adminOrReputationOracle {
         require(
             status == EscrowStatuses.Paid || status == EscrowStatuses.Partial,
```

Comment:
```suggestion
            status == EscrowStatuses.Paid || status == EscrowStatuses.Partial || status === EscrowStatuses.ToCancel
```

Reply:
What I can suggest as an alternative is to change `cancel` method signature to:
1) if status is not `toCancled` - move it to `ToCancel` (so on the first call we request cancellation)
2) if status is already `toCancel` - we check for necessary conditions to be met (e.g. refund made and etc) and if all good - move it to `Cancelled` explicitly

Reply:
I think it solves it https://github.com/humanprotocol/human-protocol/commit/731c16eaf5f5c8a34c4971ed9ff4f0072ed7bf86

Reply:
If it is not expired the cancellatio should be completed after all payouts haven been processed (reservedFunds = 0)

Reply:
1. cancel()
2. 2 options:
  a. call storeResults and reserve more funds if necessary
  b. do not reserve more funds.
3. bulkPayout() - until reserved funds = 0, then it will turn to Cancelled. If still the escrow holds funds these will be sent back to launcher.

Reply:
Clarified async: it's expected that `storeResults` is called after cancellation request, so `remainingFuns` becomes equal to `reservedFunds` and when `bulkPayOut` is called it will run payouts, `remainingFuns` becomes 0 and escrow will be automatically completed; we just need to make sure that we make it clear in docs